File: src/ddsstv/DDSSTVDiscriminator.m
//
//  DDSSTVDiscriminator.m
//  SSTV
//
//  Created by Robert Quattlebaum on 11/18/13.
//  Copyright (c) 2013 deepdarc. All rights reserved.
//

#import "DDSSTVDiscriminator.h"


#define kNumberRecordBuffers	4

@implementation DDSSTVDiscriminator

static bool
discriminator_output_func(void* context, const int16_t* freqs, size_t count)
{
	DDSSTVDiscriminator* self = (__bridge DDSSTVDiscriminator*)context;
	return [self.delegate feedFrequencies:freqs count:count];
}

- (id)init
{
    self = [super init];
    if (self) {
		queue = dispatch_queue_create("discriminator_queue", 0);
        discriminator = calloc(1,sizeof(*discriminator));
		ddsstv_discriminator_init(discriminator, 8000, DDSSTV_INTERNAL_SAMPLE_RATE);
		discriminator->output_func_context = (__bridge void *)(self);
		discriminator->output_func = &discriminator_output_func;
    }
    return self;
}

static void MyInputBufferHandler(	void *                          inUserData,
									AudioQueueRef                   inAQ,
									AudioQueueBufferRef             inBuffer,
									const AudioTimeStamp *          inStartTime,
									UInt32							inNumPackets,
									const AudioStreamPacketDescription *inPacketDesc)
{
	DDSSTVDiscriminator* self = (__bridge DDSSTVDiscriminator*)inUserData;
	[self feedSamples:inBuffer->mAudioData count:inNumPackets];
	AudioQueueEnqueueBuffer(inAQ, inBuffer, 0, NULL);
}

// ____________________________________________________________________________________
// Determine the size, in bytes, of a buffer necessary to represent the supplied number
// of seconds of audio data.
static int MyComputeRecordBufferSize(const AudioStreamBasicDescription *format, AudioQueueRef queue, float seconds)
{
	int packets, frames, bytes;

	frames = (int)ceil(seconds * format->mSampleRate);

	if (format->mBytesPerFrame > 0)
		bytes = frames * format->mBytesPerFrame;
	else {
		UInt32 maxPacketSize;
		if (format->mBytesPerPacket > 0)
			maxPacketSize = format->mBytesPerPacket;	// constant packet size
		else {
			UInt32 propertySize = sizeof(maxPacketSize);
			AudioQueueGetProperty(queue, kAudioConverterPropertyMaximumOutputPacketSize, &maxPacketSize,
				&propertySize);
		}
		if (format->mFramesPerPacket > 0)
			packets = frames / format->mFramesPerPacket;
		else
			packets = frames;	// worst-case scenario: 1 frame in a packet
		if (packets == 0)		// sanity check
			packets = 1;
		bytes = packets * maxPacketSize;
	}
	return bytes;
}

-(AudioQueueRef)audioQueue
{
	if(!audioQueue) {
		OSStatus status;
		AudioStreamBasicDescription recordFormat = {
			.mSampleRate = self.sampleRate,
			.mFormatID = kAudioFormatLinearPCM,
			.mFormatFlags = kAudioFormatFlagIsFloat,
			.mChannelsPerFrame = 1,
			.mBitsPerChannel = 32,
		};
		recordFormat.mBytesPerPacket = recordFormat.mBytesPerFrame =
				(recordFormat.mBitsPerChannel / 8) * recordFormat.mChannelsPerFrame;
		recordFormat.mFramesPerPacket = 1;
		status = AudioQueueNewInput(
			&recordFormat,
			MyInputBufferHandler,
			(__bridge void *)(self) /* userData */,
			NULL /* run loop */,
			NULL /* run loop mode */,
			0 /* flags */,
			&audioQueue
		);
		if(status) {
			NSLog(@"AudioQueueNewInput %d",status);
			return NULL;
		}

		// allocate and enqueue buffers
		int i,bufferByteSize = MyComputeRecordBufferSize(&recordFormat, audioQueue, 0.25);
		for (i = 0; i < kNumberRecordBuffers; ++i) {
			AudioQueueBufferRef buffer;
			AudioQueueAllocateBuffer(audioQueue, bufferByteSize, &buffer);
			AudioQueueEnqueueBuffer(audioQueue, buffer, 0, NULL);
		}
	}
	return audioQueue;
}

-(void)dealloc {
	if(audioQueue)
		CFRelease(audioQueue);
	ddsstv_discriminator_finalize(discriminator);
	free(discriminator);
}

-(BOOL)hiRes
{
	return false;
//	return ddsstv_discriminator_get_high_res(discriminator);
}

-(void)setHiRes:(BOOL)hiRes
{
//	dispatch_async(queue, ^{
//		ddsstv_discriminator_set_high_res(discriminator, hiRes);
//	});
}

-(double)sampleRate {
	return discriminator->resampler.ingest_sample_rate;
}

-(void)setSampleRate:(double)sampleRate {
	if(sampleRate != discriminator->resampler.ingest_sample_rate) {
		dispatch_async(queue, ^{
			ddsstv_discriminator_finalize(discriminator);
			ddsstv_discriminator_init(discriminator, sampleRate, DDSSTV_INTERNAL_SAMPLE_RATE);
			discriminator->output_func_context = (__bridge void *)(self);
			discriminator->output_func = &discriminator_output_func;
		});
	}
}

-(bool)feedSamples:(const float*)samples count:(NSInteger)count
{
	__block bool ret;
	dispatch_sync(queue, ^{
		ret = ddsstv_discriminator_ingest_samps(discriminator, samples, count);
	});
	return ret;
}

-(bool)feedSamples:(NSData*)samples
{
	return [self feedSamples:(const float*)[samples bytes] count:[samples length]/sizeof(float)];
}

@end

File: src/ddsstv/ddsstv-discriminator.c
//
//  ddsstv-discriminator.c
//  SSTV
//
//  Created by Robert Quattlebaum on 11/20/13.
//  Copyright (c) 2013 deepdarc. All rights reserved.
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dddsp/dddsp.h>
#include "ddsstv-discriminator.h"

static bool _dddsp_resampler_output_func(void* context, const float* samples, size_t count);

void
ddsstv_discriminator_init(ddsstv_discriminator_t discriminator, double ingest_sample_rate, double output_sample_rate)
{
	memset(discriminator,0,sizeof(*discriminator));

	dddsp_resampler_init(&discriminator->resampler, ingest_sample_rate, output_sample_rate);
	discriminator->resampler.output_func = &_dddsp_resampler_output_func;
	discriminator->resampler.output_func_context = discriminator;

	// TODO: REMOVE ME
	//discriminator->pass_thru = true;

	discriminator->ingest_processing_int16 = calloc(1,sizeof(*discriminator->ingest_processing_int16)*output_sample_rate);

	bool high_quality = true;
	bool aggressive_filtering = true;

	//high_quality = false;
	//aggressive_filtering = false;

	discriminator->freq_disc = dddsp_discriminator_alloc(
		discriminator->resampler.output_sample_rate,
		1700.0/discriminator->resampler.output_sample_rate,
		0,
		//(high_quality&&aggressive_filtering)?600.0/discriminator->resampler.output_sample_rate:0,
		high_quality
	);

	if (high_quality) {
		discriminator->freq_disc_sync = dddsp_discriminator_alloc(
			discriminator->resampler.output_sample_rate,
			1200.0/discriminator->resampler.output_sample_rate,
			300.0/discriminator->resampler.output_sample_rate,
			high_quality
		);
	}

	if (aggressive_filtering) {
		discriminator->filter = dddsp_fir_float_alloc_band_pass(
			1000.0/discriminator->resampler.output_sample_rate,
			2500.0/discriminator->resampler.output_sample_rate,
			11,
			DDDSP_BLACKMAN
		);
    }
}

void
ddsstv_discriminator_finalize(ddsstv_discriminator_t discriminator)
{
	dddsp_resampler_finalize(&discriminator->resampler);
	free(discriminator->ingest_processing_int16);
	memset(discriminator,0,sizeof(*discriminator));
}

bool
_dddsp_resampler_output_func(void* context, const float* samples, size_t count)
{
	ddsstv_discriminator_t discriminator = context;
	bool ret = false;
	ssize_t freq_count = 0;
	int16_t* freq_buffer = discriminator->ingest_processing_int16;

	if (!count) {
		ret = discriminator->output_func(
			discriminator->output_func_context,
			NULL,
			0
		);
	}

	while (count--) {
		float v = *samples++;
		v = dddsp_iir_float_feed(discriminator->filter, v);
		if (!discriminator->pass_thru) {
			if (discriminator->freq_disc_sync) {
				float low_v = dddsp_discriminator_feed(discriminator->freq_disc_sync, v);
				v = dddsp_discriminator_feed(discriminator->freq_disc, v);
				if (v < 0) {
					v = low_v;
				}
				else
				if (!isfinite(v) || (low_v<1350.0 && v<1400.0 && low_v>1000.0)) {
					v = low_v;
				}
				else if (dddsp_discriminator_get_last_magnitude(discriminator->freq_disc_sync) > dddsp_discriminator_get_last_magnitude(discriminator->freq_disc)) {
					v = low_v;
				}
			} else {
				v = dddsp_discriminator_feed(discriminator->freq_disc, v);
			}

			v = dddsp_clampf(v, 0.0, 0.0, 2800.0);
			v = dddsp_median_float_feed(&discriminator->median_filter, v);
		}

		// Quantize
		freq_buffer[freq_count++] = (int16_t)(v+0.5f);

		if (freq_count >= discriminator->resampler.output_sample_rate) {
			ret |= discriminator->output_func(
				discriminator->output_func_context,
				freq_buffer,freq_count
			);
			freq_count = 0;
		}
	}
	if(freq_count) {
		ret |= discriminator->output_func(
			discriminator->output_func_context,
			freq_buffer,freq_count
		);
	}
	return ret;
}


bool
ddsstv_discriminator_ingest_samps(ddsstv_discriminator_t discriminator, const float* samps, size_t count)
{
	return dddsp_resampler_ingest_samps(&discriminator->resampler, samps, count);
}

//bool
//ddsstv_discriminator_get_high_res(ddsstv_discriminator_t discriminator)
//{
//	return discriminator->hires_filter != NULL;
//}
//
//void
//ddsstv_discriminator_set_high_res(ddsstv_discriminator_t discriminator, bool hi_res)
//{
//	float filter_top = 2350.0;
//	float filter_bottom = 1050.0;
//
//	if(hi_res==ddsstv_discriminator_get_high_res(discriminator))
//		return;
//
//	if(hi_res) {
//		discriminator->hires_filter = dddsp_iir_float_alloc_band_pass(
//			filter_bottom/discriminator->ingest_sample_rate*2,
//			filter_top/discriminator->ingest_sample_rate*2,
//			0.1,
//			4
//		);
//	} else {
//		dddsp_iir_float_finalize(discriminator->hires_filter);
//		discriminator->hires_filter = NULL;
//	}
//}
